crates/table/src/btree_index.rs

```diff
@@ -186,13 +186,57 @@ impl TypedIndex {
             Self::String(this) => insert_at_type(this, cols, row_ref),
 
             Self::AlgebraicValue(this) => {
-                let key = row_ref.project_not_empty(cols)?;
+                let key = row_ref.project(cols)?;
                 this.insert(key, row_ref.pointer());
                 Ok(())
             }
         }
     }
 
+    /// Add the row referred to by `row_ref` to the index `self`,
+    /// which must be keyed at `cols`.
+    ///
+    /// If `cols` is inconsistent with `self`,
+    /// or the `row_ref` has a row type other than that used for `self`,
+    /// this will behave oddly; it may return an error,
+    /// or may insert a nonsense value into the index.
+    /// Note, however, that it will not invoke undefined behavior.
+    ///
+    /// Assumes that this is a unique constraint and returns `Ok(Some(existing_row))` if it's violated.
```

Comment:
Add a line saying whether the index is mutated in that case.

I don't love this return type: it seems like a unique constraint violation is error-ey. I would consider `Result<Result<(), RowPointer>, InvalidFieldError>` or defining `enum InsertUniqueError { InvalidField(InvalidFieldError), AlreadyPresent(RowPointer) }` and returning `Result<(), InsertUniqueError>`. But this is a small nit and it's fine with me if we merge with the type as-is. I do care about the comment, though.